core/ledger/kvledger/txmgmt/statedb/statecouchdb/ccupgradehandler.go

```diff
@@ -0,0 +1,29 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package statecouchdb
+
+import (
+	"fmt"
+	"github.com/hyperledger/fabric/extensions/gossip/api"
+	"strings"
+)
+
+const lsccKeyPrefix = "%s~"
+
+//getCCUpgradeHandler returns block publisher chaincode upgrade handler related to given versioned DB instance
+func getCCUpgradeHandler(vdb *VersionedDB) api.ChaincodeUpgradeHandler {
+	return func(txMetadata api.TxMetadata, chaincodeName string) error {
+		logger.Debugf("Clearing lscc state cache for chaincode [%s]", chaincodeName)
+
+		for k := range vdb.lsccStateCache.cache {
```

Comment:
This isn't thread-safe.

Reply:
vdb.lsccStateCache.evictEntry(k) is lock based.

https://github.com/trustbloc/fabric-mod/pull/110/files#diff-d3b1a6c4a94dc2ad324f2ce6e812f77bR140

Reply:
> vdb.lsccStateCache.evictEntry(k) is lock based.
> 
> https://github.com/trustbloc/fabric-mod/pull/110/files#diff-d3b1a6c4a94dc2ad324f2ce6e812f77bR140

But you're iterating over the map outside of the lock.

Reply:
Changed evictEntry logic to make it threadsafe